### ISSUE TITLE
visual studio code extension needs references/definitions

### ISSUE THREAD

**Issue by seanyoung** [hyperledger/solang#714]
The visual studio code extension gives compiler diagnostics (i.e. compiler warnings and errors). For functions, variables, and all other types (events, contracts, etc) it should be able to:
- Go to definition
- Go to type definition
- Find references
- Find all references
- Rename (and rename all the references)

**Comment by gitcoinbot** [comment]
Issue Status: 1. **Open** 2. Started 3. Submitted 4. Done 

<hr>

__This issue now has a funding of 600.0 DAI (600.0 USD @ $1.0/DAI)  attached to it.__

 * If you would like to work on this issue you can 'start work' [on the Gitcoin Issue Details page](https://gitcoin.co/issue/hyperledger-labs/solang/714/100028726).
* Want to chip in? Add your own contribution [here](https://gitcoin.co/issue/hyperledger-labs/solang/714/100028726).
* Questions? Checkout <a href='https://gitcoin.co/help'>Gitcoin Help</a> or the <a href='https://discord.gg/gitcoin/'>Gitcoin's Discord</a>
* $4,682,892.87 more funded OSS Work available on the [Gitcoin Issue Explorer](https://gitcoin.co/explorer)

**Comment by gitcoinbot** [comment]
Issue Status: 1. **Open** 2. Started 3. Submitted 4. Done 

<hr>

__Workers have applied to start work__.


These users each claimed they can complete the work by 264 years, 7 months from now.
Please review their action plans below:


**1) [vividwood](https://gitcoin.co/vividwood) has applied to start work _(Funders only: [approve worker](https://gitcoin.co/issue/hyperledger-labs/solang/714/100028726?mutate_worker_action=approve&worker=vividwood) | [reject worker](https://gitcoin.co/issue/hyperledger-labs/solang/714/100028726?mutate_worker_action=reject&worker=vividwood))_.**

Hi, I have experience in developing VS Code extensions and will be glad to work on this bounty!

Learn more [on the Gitcoin Issue Details page](https://gitcoin.co/issue/hyperledger-labs/solang/714/100028726).

**Comment by seanyoung** [comment]
@Web3Foundation can you please approve @theboywhomakesgames, thank you.

**Comment by bshevchenko** [comment]
Got it 👌🏻

**Comment by bshevchenko** [comment]
@Web3Foundation @seanyoung got it 👌🏻

**Comment by seanyoung** [comment]
@bshevchenko have you made any progress on this bounty?

**Comment by littlezigy** [comment]
Hello
Is this bounty still open?

**Comment by seanyoung** [comment]
@littlezigy noone is working on this so the bounty is still available. If you would like to work on it, please do. While you are working on it apply on the gitcoin page too, thanks.

**Comment by littlezigy** [comment]
Nice! I'll be awaiting your approval.

**Comment by seanyoung** [comment]
Just before we approve you - what's the plan to do this work? Do you have any rust experience?

**Comment by littlezigy** [comment]
I have only a little experience with rust, but I have a lot of experience with Go. However, I pick up languages very fast.

I plan to add the following functionality described in the project description

- Go to definition
- Find (all) references
- Rename (all) references

I will add tests for the new features to (the test folder).
I assume this is the extension directory: https://github.com/hyperledger-labs/solang/tree/main/vscode

**Comment by seanyoung** [comment]
> I will add tests for the new features to (the test folder). I assume this is the extension directory: https://github.com/hyperledger-labs/solang/tree/main/vscode

The language server is written in rust, not typescript: https://github.com/hyperledger-labs/solang/blob/main/src/bin/languageserver/mod.rs

**Comment by seanyoung** [comment]
@Web3Foundation can you please approve @littlezigy, thank you.

**Comment by littlezigy** [comment]
Ah okay. Thanks

**Comment by littlezigy** [comment]
Hi @seanyoung. Is it okay if I edit the extension so it uses docker solang if it is available?

**Comment by seanyoung** [comment]
> Hi @seanyoung. Is it okay if I edit the extension so it uses docker solang if it is available?

Why would you want it to use docker? This introduces another dependency, which really needs justification.

**Comment by littlezigy** [comment]
> > Hi @seanyoung. Is it okay if I edit the extension so it uses docker solang if it is available?
> 
> Why would you want it to use docker? This introduces another dependency, which really needs justification.

The binary doesn't work on my computer (Ubuntu 18.04) so I'm using docker. The extension keeps downloading and calling the binary package. I still plan to keep it working this way, but I want to add a check for the docker image then have the extension use that if the user sets that in the extension settings.

**Comment by littlezigy** [comment]
> Hi @seanyoung. Is it okay if I edit the extension so it uses docker solang if it is available?

I won't be doing this anymore. I figured out a way to handle this.

**Comment by littlezigy** [comment]
@Web3Foundation  could you approve me?

**Comment by darkdrag00n** [comment]
Hi @seanyoung, I am interested in working on this and have applied on the Gitcoin bounty. I've shared my plan on Gitcoin but also sharing it here for reference.

I have worked professionally in Rust for 1 year and have built toy language compilers in my free time.

I've looked at the LanguageServer (`src/bin/languageserver`) and the AST definitions briefly. I believe most of the information needed to serve these client queries should be present in the `sema::ast::Namespace` object which is returned when the server parses and resolves the file.

I estimate that I'll take ~2-3 weeks post approval to get the PR ready for review. Post that we can iterate on the suggestions to improve the code further. I've already joined the #solang channel on Discord (darkdrag00n#6796) and might reach out for queries/updates if needed.

Thanks!
